### PR TITLE
Change Display Name to Functional Component

### PR DESCRIPTION
**Pull request by srikarparsi**
### Fixed Issues

$ https://github.com/Expensify/App/issues/16141

### Tests

1. Open New Dot
2. Ensure that the display names are shown properly
3. On web and desktop, ensure that hovering over the name displays a tooltip with the right data.

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

image

</details>

<details>
<summary>Mobile Web - Safari</summary>

image

</details>

<details>
<summary>Desktop</summary>

image

</details>

<details>
<summary>iOS</summary>

Simulator Screen Shot - iPhone 14 Pro - 2023-08-01 at 10 16 19

</details>

<details>
<summary>Android</summary>

image

</details>

### CODE REVIEW

**Comment by srikarparsi** [comment]
Hi @aimane-chnaif, I just added it as a draft to finish working on later but I should get this ready to review today and will ping you once I do.

**Comment by srikarparsi** [comment]
Hi @gedu! I tried addressing the feedback in this [comment](https://github.com/Expensify/App/issues/16141#issuecomment-1620756970) but didn't completely understand one thing. Why can't `getTooltipShiftX(index)` be called inside a lamda function as `shiftHorizontal={() => getTooltipShiftX(index)}`? I may have missed some other things as well as I am still learning about React Hooks so please let me know if I have missed any other changes.

**Comment by srikarparsi** [comment]
@aimane-chnaif @gedu, can we make this useCallback have an empty dependency array? containerLayout and childRefs don't change except when the component is mounted so not sure if we have to include them in the dependency array.

**Comment by gedu** [comment]
> Hi @gedu! I tried addressing the feedback in this https://github.com/Expensify/App/issues/16141#issuecomment-1620756970 but didn't completely understand one thing. Why can't getTooltipShiftX(index) be called inside a lamda function as shiftHorizontal={() => getTooltipShiftX(index)}? I may have missed some other things as well as I am still learning about React Hooks so please let me know if I have missed any other changes.

Because the component `TooltipRenderedOnPageBody` uses that props as dependency of the `useMemo` and that lamda isn't a stable function, it gets recreated after every frame, so will re-executed the hook

**Comment by srikarparsi** [comment]
Hi @gedu! I tried finding a couple of ways to avoid using the lambda function but I couldn't find any good ones. The only option I could come find is `getTooltipShiftX.bind(null, index)` but I believe this has the same problem of copying the [function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Function/bind#:~:text=A%20copy%20of%20the%20given%20function%20with%20the%20specified%20this%20value%2C%20and%20initial%20arguments%20(if%20provided).). I'm not sure if I'm overlooking something simple but would you recommend an approach to avoid the lambda?

**Comment by gedu** [comment]
> Hi @gedu! I tried finding a couple of ways to avoid using the lambda function but I couldn't find any good ones. The only option I could come find is `getTooltipShiftX.bind(null, index)` but I believe this has the same problem of copying the [function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_objects/Function/bind#:~:text=A%20copy%20of%20the%20given%20function%20with%20the%20specified%20this%20value%2C%20and%20initial%20arguments%20(if%20provided).). I'm not sure if I'm overlooking something simple but would you recommend an approach to avoid the lambda?

Yes, first I would split the component because if `tooltipEnabled` is false we don't need any state, callbacks, etc.

```
function DisplayNames(props) {
    if (!props.tooltipEnabled) {
        return (
            <DisplayNamesWithoutTooltip
                textStyles={props.textStyles}
                numberOfLines={props.numberOfLines}
                fullTitle={props.fullTitle}
            />
        );
    }

    // eslint-disable-next-line react/jsx-props-no-spreading
    return <DisplayNamesTooltip {...props} />;
}
```
Then in the `_.map` I would create a new Component to handle better this lambda function

```
function UserTooltipItem(props) {
    const tooltipIndexBridge = useCallback(() => props.getTooltipShiftX(index), [props.getTooltipShiftX, props.index]);

    return (
        <Fragment key={props.index}>
            <UserDetailsTooltip
                // props
                shiftHorizontal={tooltipIndexBridge}
            >
            // All the code from _.map
 }
```
And you use it like:            

```
_.map(props.displayNamesWithTooltips, ({displayName, accountID, avatar, login}, index) => (
                      <UserTooltipItem key={index} getTooltipShiftX={getTooltipShiftX} //pass the props you need />
))}
```

Hope this makes sense

**Comment by srikarparsi** [comment]
Thanks for the tip @gedu! I went ahead with your advice and created a bridge function within the map so that we're not re-creating the getTooltipShiftX function on multiple renders. I did keep it all in the same component though since I didn't think it was worth creating a new UserToolTipItem component to handle this functionality.

I also do agree that we should split the component based on tooltipEnabled but I think this should be done in a different PR as I think it would make sense to focus on just the refactor to a functional component.

Do you think you could take a look at this whenever you get a chance? @gedu @aimane-chnaif

**Comment by gedu** [comment]
@srikarparsi you have a Lint error, you can't have `useCallback` inside of a callback

**Comment by gedu** [comment]
> I also do agree that we should split the component based on tooltipEnabled but I think this should be done in a different PR

Just notice this, In my opinion, there is no better time than now while refactoring the component, you don't even have to create a new file, can be in the same file, but well I think could be @aimane-chnaif call

**Comment by srikarparsi** [comment]
> Just notice this, In my opinion, there is no better time than now while refactoring the component, you don't even have to create a new file, can be in the same file, but well I think could be @aimane-chnaif call

Yeah that is true I agree, I'll ask internally to make sure that refactoring components as a whole can be done with refactoring to functional components as we usually like to keep PRs focussed and will make the change if all is okay.

**Review by aimane-chnaif — COMMENTED** [review]
There are several lint errors to fix

https://github.com/Expensify/App/assets/96077027/a9170d89-1b08-4800-a60b-e8b958547c49

**Comment by srikarparsi** [comment]
Hi @aimane-chnaif, yeah I changed the title to "WIP (work in progress)" because I was working on some other bugs but I'll fix these and re-request review tomorrow or the day after.

**Comment by srikarparsi** [comment]
This is ready for re-review @aimane-chnaif @gedu

File: src/components/DisplayNames/DisplayNamesWithTooltip.js [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        // No need for any complex text-splitting, just return a simple Text component
+        return (
+            <Text
+                style={[...props.textStyles, props.numberOfLines === 1 ? styles.pre : styles.preWrap]}
+                numberOfLines={props.numberOfLines}
+            >
+                {props.fullTitle}
+            </Text>
+        );
+    }
```
Comment (gedu) [inline src/components/DisplayNames/DisplayNamesWithTooltip.js:62]: is this duplicated? I think this case is being handle at `index` file

File: src/components/DisplayNames/DisplayNamesWithoutTooltip.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    numberOfLines: PropTypes.number,
+};
+
+const defaultProps = {
+    fullTitle: '',
+    textStyles: [],
+    numberOfLines: 1,
+};
+
+function DisplayNamesWithoutTooltip(props) {
```
Comment (gedu) [inline src/components/DisplayNames/DisplayNamesWithoutTooltip.js:23]: you can use destruction

```
function DisplayNamesWithoutTooltip({textStyles, numberOfLines, fullTitle}) {}
```

File: src/components/DisplayNames/DisplayNamesWithTooltip.js (hunk 1 above)
Comment (srikarparsi) [reply src/components/DisplayNames/DisplayNamesWithTooltip.js]: You're right, forgot to remove that thank you

**Comment by srikarparsi** [comment]
Thanks @gedu, just made the changes and re-uploaded all screenshots.

**Comment by srikarparsi** [comment]
Hi @aimane-chnaif, can you please take a look at this whenever you get a chance

**Comment by srikarparsi** [comment]
bump @aimane-chnaif

**Comment by srikarparsi** [comment]
Hi @aimane-chnaif, did you have a chance to look at this. If you're busy, please let me know and I can ask if there are other C+s who have a spare cycle.

**Review by aimane-chnaif — COMMENTED** [review]
Looks good. Just minor feedback:

Also, please pull main

**Comment by srikarparsi** [comment]
Thanks @aimane-chnaif, made those changes so this is ready for re-review.

**Review by aimane-chnaif — APPROVED** [review]
LGTM 🎉

**Review by PauloGasparSv — APPROVED** [review]
LGTM, tested on Web too and is working fine!

File: src/components/DisplayNames/DisplayNamesWithTooltip.js [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+function DisplayNamesWithToolTip(props) {
+    const containerRef = useRef(null);
+    const childRefs = useRef([]);
+    const [isEllipsisActive, setIsEllipsisActive] = useState(false);
+
+    useEffect(() => {
+        setIsEllipsisActive(
+            containerRef.current && containerRef.current.offsetWidth && containerRef.current.scrollWidth && containerRef.current.offsetWidth < containerRef.current.scrollWidth,
+        );
+    }, []);
```
Comment (abdulrahuman5196) [inline src/components/DisplayNames/DisplayNamesWithTooltip.js:19]: Ellipsis is not updated when the `fullTitle` prop is changed which is lead to this issue - https://github.com/Expensify/App/issues/27545

**Pull request merged**